### PR TITLE
Make Moment immutable --> dependent PR--> #2378

### PR DESCRIPTION
**Pull request by jitendrs**
Raise new pull requests since old requests got a lot of unwanted code changes.
manual merge for conflict files. 
Please review the code and approve PR.
Fix license issue.
Fix merge issues occurred during the build.
Injected require code changes for momment.py

### CODE REVIEW

**Comment by jitendrs** [comment]
@vtomole Which once you are referring too? Can you provide more details

File: cirq/ops/moment.py [hunk 1]
```diff
@@ -110,10 +113,10 @@ def without_operations_touching(self, qubits: Iterable[raw_types.Qid]):
     def __copy__(self):
         return type(self)(self.operations)
 
-    def __bool__(self) -> bool:
+    def __bool__(self):
```
Comment (vtomole) [inline cirq/ops/moment.py]: Master has 
`def __bool__(self) -> bool:`

File: cirq/ops/moment.py [hunk 2]
```diff
@@ -110,10 +113,10 @@ def without_operations_touching(self, qubits: Iterable[raw_types.Qid]):
...
         return bool(self.operations)
 
-    def __eq__(self, other) -> bool:
+    def __eq__(self, other):
```
Comment (vtomole) [inline cirq/ops/moment.py]: ` def __eq__(self, other) -> bool:`

File: cirq/ops/moment.py [hunk 3]
```diff
@@ -131,14 +134,14 @@ def _approx_eq_(self, other: Any, atol: Union[int, float]) -> bool:
                                           key=lambda op: op.qubits),
                                    atol=atol)
 
-    def __ne__(self, other) -> bool:
+    def __ne__(self, other):
```
Comment (vtomole) [inline cirq/ops/moment.py]: `def __ne__(self, other) -> bool:`

File: cirq/ops/moment.py [hunk 4]
```diff
@@ -131,14 +134,14 @@ def _approx_eq_(self, other: Any, atol: Union[int, float]) -> bool:
...
         return not self == other
 
     def __hash__(self):
         return hash(
             (Moment, tuple(sorted(self.operations, key=lambda op: op.qubits))))
 
-    def __iter__(self) -> Iterator['cirq.Operation']:
+    def __iter__(self):
```
Comment (vtomole) [inline cirq/ops/moment.py]: `def __iter__(self) -> Iterator['cirq.Operation']:`

File: cirq/ops/moment.py (hunk 1 above)
Comment (jitendrs) [reply cirq/ops/moment.py]: ya I gotcha.

**Comment by jitendrs** [comment]
Please review this PR and consequently close dependent PR's.

**Comment by jitendrs** [comment]
@vtomole Please approve PR as all required changes are implemented. 
Thanks for helping me.

File: cirq/ops/moment_test.py [hunk 5]
```diff
@@ -253,3 +253,11 @@ def test_inverse():
...
     assert cirq.inverse(cirq.inverse(m)) == m
     assert cirq.inverse(cirq.Moment([cirq.measure(a)]), default=None) is None
+
+
+def test_immutable_moment():
+    with pytest.raises(AttributeError):
+        q1, q2 = cirq.LineQubit.range(2)
+        circuit = cirq.Circuit(cirq.X(q1))
+        moment = circuit.moments[0]
+        moment.operations += (cirq.Y(q2),)
```
Comment (vtomole) [inline cirq/ops/moment_test.py]: Add a newline.

Comment (jitendrs) [reply cirq/ops/moment_test.py]: Where do you want me to add a line, at the end of the program?

Comment (vtomole) [reply cirq/ops/moment_test.py]: Yes.

Comment (jitendrs) [reply cirq/ops/moment_test.py]: I see. Thanks.

**Comment by jitendrs** [comment]
Thanks to you for pointing a needle in the grass.

**Pull request merged**